### PR TITLE
feat(streaming): only output required columns for MV-on-MV

### LINKED ISSUE #4529: streaming: column pruning in dispatcher for MV on MV
The Chain executor is actually compiled from `LogicalScan`, which may only require a subset of columns from the upstream materialized view. For now, the dispatcher at the upstream mview will dispatch the chunk verbatim to the chain and let the chain prune the columns.

Considering that the scheduler always puts the chain executor to the same parallel unit as the upstream materialize, the outputs of the dispatchers must be local ones so there's not too much overhead. However, it's possible we put the chain to a remote compute node. Allow column pruning in the dispatcher may significantly reduce the exchange I/O.

### PR DESCRIPTION
**Pull request by BugenZhao**
Signed-off-by: Bugen Zhao <i@bugenzhao.com>I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

The sequel to #8533. We'll now have the same schema for the upstream `Merge` and the snapshot (or table scan). See the PR body of #8533 for more details.

- Related: #8351
- Close #4529 

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/meta/src/stream/stream_graph/fragment.rs [hunk 1]
```diff
@@ -52,6 +52,9 @@ pub(super) struct BuildingFragment {
 
     /// The ID of the job if it's materialized in this fragment.
     table_id: Option<u32>,
+
+    /// The required columns of each upstream table.
+    upstream_table_columns: HashMap<TableId, Vec<i32>>,
```
Comment (chenzl25) [inline src/meta/src/stream/stream_graph/fragment.rs:57]: Is it possible for a fragment to contain more than one upstream table?

**Comment by BugenZhao** [comment]
After this PR, if a column is not referenced in all of the downstream materialized views, we can freely drop the column with `ALTER TABLE DROP`. 🤓

File: src/meta/src/stream/stream_graph/fragment.rs [hunk 2]
```diff
@@ -138,6 +143,28 @@ impl BuildingFragment {
...
+    ) -> HashMap<TableId, Vec<i32>> {
+        let mut table_columns = HashMap::new();
+
+        visit::visit_fragment(fragment, |node_body| {
+            if let NodeBody::Chain(chain_node) = node_body {
+                let table_id = chain_node.table_id.into();
+                let column_ids = chain_node.upstream_column_ids.clone();
+                table_columns
+                    .try_insert(table_id, column_ids)
+                    .expect("currently there should be no two same upstream tables in a fragment");
```
Comment (chenzl25) [inline src/meta/src/stream/stream_graph/fragment.rs:160]: L160 asserts there should be no two same upstream tables in a fragment, but is it possible to have two different upstream tables?

Comment (BugenZhao) [reply src/meta/src/stream/stream_graph/fragment.rs:160]: I think it's currently impossible. The reason for why using a collection here is explained in https://github.com/risingwavelabs/risingwave/pull/4740#discussion_r949840701.

**Review by chenzl25 — APPROVED** [review]
LGTM!

File: src/meta/src/stream/stream_graph/fragment.rs (hunk 2 above)
Comment (chenzl25) [reply src/meta/src/stream/stream_graph/fragment.rs:160]: Thanks for your explanation.

**Pull request merged**